RaftMongoWithRaftReconfig.tla

```diff
@@ -0,0 +1,276 @@
+    /\
+       \/ Len(log[i]) > Len(log[j])
+       \* There seems no short-cut of OR clauses, so I have to specify the negative case
+       \/ /\ Len(log[i]) <= Len(log[j])
+          /\ LastTerm(log[i]) /= LogTerm(j, Len(log[i]))
+
+RollbackOplog(i, j) ==
+    /\ CanRollbackOplog(i, j)
+    \* Rollback 1 oplog entry
+    /\ LET new == [index2 \in 1..(Len(log[i]) - 1) |-> log[i][index2]]
+         IN log' = [log EXCEPT ![i] = new]
+    /\ UNCHANGED <<serverVars, configs>>
+
+\* The set of nodes that has log[me][logIndex] in their oplog
+Agree(me, logIndex) ==
+    { node \in ServerViewOn(me) :
+        /\ Len(log[node]) >= logIndex
+        /\ LogTerm(me, logIndex) = LogTerm(node, logIndex) }
+
+NotBehind(me, j) == \/ LastTerm(log[me]) > LastTerm(log[j])
+                    \/ /\ LastTerm(log[me]) = LastTerm(log[j])
+                       /\ Len(log[me]) >= Len(log[j])
+
+\* ACTION
+\* i = the new primary node.
+BecomePrimaryByMagic(i, ayeVoters) ==
+    /\ \A j \in ayeVoters : /\ i \in ServerViewOn(j)
+                            /\ NotBehind(i, j)
+                            /\ currentTerm[j] <= currentTerm[i]
+    /\ ayeVoters \in Quorum(i)
+    /\ state' = [index \in Server |-> IF index \notin ayeVoters
+                                      THEN state[index]
+                                      ELSE IF index = i THEN Leader ELSE Follower]
+    /\ currentTerm' = [index \in Server |-> IF \/ index \in ayeVoters 
+                                               \/ index = i
+                                            THEN currentTerm[i] + 1 
+                                            ELSE currentTerm[index]]
+    /\ LET noop == [term  |-> currentTerm[i] + 1, configVersion |-> GetConfigVersion(i)]
+           newLog == Append(log[i], noop)
+       IN  log' = [log EXCEPT ![i] = newLog]
+    /\ UNCHANGED <<commitPoint, configs>>
+
+\* ACTION
+\* Leader i receives a client request to add v to the log.
+ClientWrite(i) ==
+    /\ state[i] = Leader
+    /\ LET entry == [term  |-> currentTerm[i], configVersion |-> GetConfigVersion(i)]
+           newLog == Append(log[i], entry)
+       IN  log' = [log EXCEPT ![i] = newLog]
+    /\ UNCHANGED <<serverVars, configs>>
+        
+\* ACTION
+AdvanceCommitPoint ==
+    \E leader \in Server : \E acknowledgers \in SUBSET Server :
+        /\ state[leader] = Leader
+        /\ acknowledgers \subseteq Agree(leader, Len(log[leader]))
+        /\ acknowledgers \in Quorum(leader)
+        \* If we comment out the following line, a replicated log entry from old primary will voilate the safety.
+        \* [ P (2), S (), S ()]
+        \* [ S (2), S (), P (3)]
+        \* [ S (2), S (2), P (3)] !!! the log from term 2 shouldn't be considered as committed.
+        /\ LogTerm(leader, Len(log[leader])) = currentTerm[leader]
+        \* If an acknowledger has a higher term, the leader would step down.
+        /\ \A j \in acknowledgers : currentTerm[j] <= currentTerm[leader]
```

Comment:
I think this is implied by line 171 and line 165.

Reply:
Without this line, we trigger NeverRollbackCommitted. For example:

Node A becomes primary in term 1 and writes a noop
A: [1]
B:
C:

Node B becomes primary in term two with votes from B and C and writes a noop
A: [1]
B: [2]
C:

C appends an oplog entry from A
A: [1]
B: [2]
C: [1]

A commits [1]. Now [1] can be rolled back.

Adding bi-directional term propagation in AppendOplog doesn't fix this because of chaining.

Reply:
I could have written currentTerm[j] = currentTerm[leader] since AppendOplog updated the downstream node's term. However, the <= is important if I remove the term update. An example is

A becomes primary in term 2 with votes from A and B. It writes a noop.
A: [1] [2]
B: [1]
C: [1]

A replicates the noop to node C.
A: [1] [2]
B: [1]
C: [1] [2]

A can acknowledge the noop [2] as majority committed, even though C has term=1.

Reply:
I think Line 171 is still needed, or we would incorrectly majority commit an entry from a previous term.